### ISSUE TITLE
'Reconnect' doesn't appear to re-establish connection?

### ISSUE THREAD

**Issue by nebev** [superhawk610/multicast#34]
**Context:**
We're using this awesome tool in several meeting rooms. We've got 4 Sony Bravia Android TVs with Chromecast enabled on them (all running latest software). At the start of the day, we restart multicast, and they all show a calendar. However, during the day people navigate off the chromecast mode to do things like connect their laptop (which is fine).

**Problem**
When multicast starts, it appears to reconnect all Chromecast-enabled TVs and they all display as they're supposed to. However, once they've had something else on them, they cannot 'switch back' to multicast without a restart of multicast. This isn't ideal as _all_ chromecasts refresh (which then takes over the TV). One would assume that this is what the 'reconnect' button is supposed to do, but it doesn't do anything? Ideally I'd like to reconnect a single chromecast, rather than all of them (if possible).

Am I incorrect in assuming that this is what the 'reconnect' button is supposed to do? Or is this a bug, or a quirk with Chromecast-enabled TVs?

**Comment by Brekmister** [comment]
I encounter the same behavior with all generations of Chromecast and,
Chromecast Ultra. This probably will have to be fixed.

**Comment by superhawk610** [comment]
It definitely sounds like a bug. The **Reconnect** button on the Device page shoots a GET request to `/devices/:device_id/connect`, which in turn calls `launchHub()` with the device's IP address. `launchHub()` then attempts to go through the pairing/launching process for *that specific IP address*, so it should do what you'd expect it to do, namely reconnecting that receiver.

Is it possible that the Chromecasts are being assigned a different IP address at some point throughout the day? Can you try loading up multicast and then immediately using the **Reconnect** button to see if it has any effect?

**Comment by superhawk610** [comment]
@Brekmister can you confirm that the Chromecast remain on the same IP address and fails to respond to the **Reconnect** button?

**Comment by Brekmister** [comment]
@superhawk610 Huh, It somehow worked this time on the master branch, I'll
edit this comment when I can confirm on the different scenarios

@nebev Can you tell us how multicast is setup and the layout of your
network in terms of where the multicast host is in relation to the
Chromecasts? Do you have any managed switches/firewalls/WiFi routers that
goes in between Chromecast and Multicast?

**Comment by nebev** [comment]
Hey. Can confirm that the IP addresses have not changed.
I've got both the server and all the chromecasts on a WPA-2 network. There are some switches behind the scenes to link the APs together, but can confirm that reconnect does not work even when server and chromecast are on the same AP.

What gets me is that it works flawlessly on application startup. Assuming that the code is not the same for reconnecting?

**Comment by superhawk610** [comment]
@nebev Have you tried to replicate this since v2.0? I've modified the way that Reconnect functions, let me know if this issue persists.

**Comment by nebev** [comment]
Hey.
Wasn't aware of the new version. Just pulled and am now running commit c79b4851e8a1c6ebb7913d78888dd6f78ccb2f6a (2.0.1 I believe)

I now get a notification that the client has disconnected (pretty sure that wasn't there before?) And a notification in the UI that something is happening when "Reconnect" is pressed, but alas, the issue remains.

Logs are as follows:
```
0|multicas | MultiCast is live!
0|multicas | Listening at port 3944...
0|multicas | connected client with host: 192.168.67.14
0|multicas | connected client with host: 192.168.67.108
0|multicas | connected client with host: 192.168.67.202
0|multicas | disconnected client with host: 192.168.67.108
```

No extra log lines are present when Reconnect is pressed.

**Comment by superhawk610** [comment]
Can you launch `chrome://inspect` on a Chromium browser and click 'inspect' on the device that's disconnecting?

The dev console there may give you some insight into what's happening. See if any network requests are being blocked or failing.

Does the device in question appear as online or offline in the Devices page? This could potentially be why you're not seeing anything happen when you press Reconnect, if the device status is 'offline' or 'waiting' then it won't do anything. This may be incorrect behavior on Multicast's part, I may need to fix this.

**Comment by nebev** [comment]
I assume you're talking about Chrome's remote inspector?
![image](https://user-images.githubusercontent.com/1707973/34798192-2d707710-f6af-11e7-90dc-17dc533c64c1.png)
I'm not seeing any applications on any of the devices, even when Multicast is 'on'.

For clarification, the multicast app itself on the TV is working fine on startup. And it's exiting cleanly (I believe) when someone presses "Back" or "Exit". It's just re-casting specifically to that device at another time that is an issue. What would you expect to see in the remote console if Multicast is not running?

Not sure if it's related, but I'm now seeing:
`Something's fishy.
  Receiver missed too many heartbeats, it is likely offline.`

And the disconnected device is still coming up with a green dot next to it (assuming that means Multicast believes it's still connected).

**Comment by superhawk610** [comment]
This is what I see:

![remote_devices](https://user-images.githubusercontent.com/18172185/34831794-f8f03a50-f6b5-11e7-834a-0c0bcc019b30.png)

It worries me that it's not showing that you're running anything or allowing you to Inspect the device, as that completely prevents you from debugging. It makes me think it's a TV-integrated device issue.

@GYatesIII may have some input here, he got Multicast working on a TV-integrated Chromecast a few months back. I remember he had to [manually update the firmware](https://github.com/superhawk610/multicast/issues/26#issuecomment-344104631) though, maybe something to look into.

With the next update I'll make that error message also show which device triggered the error, check back in a few days.

**Comment by superhawk610** [comment]
Closed in 4f4d516f593cd007daf6501ea9750adfd48ac081.
